components/webext-storage/src/migration.rs

```diff
@@ -178,18 +212,72 @@ fn read_rows(filename: &Path) -> Vec<LegacyRow> {
         Ok(r) => r,
         Err(e) => {
             log::warn!("Failed to read any rows from the source DB: {}", e);
-            return Vec::new();
+            return (
+                Vec::new(),
+                MigrationInfo {
+                    non_fatal_errors_seen: 1,
+                    open_failure: true,
+                    ..Default::default()
+                },
+            );
         }
     };
+    let rows: Vec<LegacyRow> = rows.filter_map(Result::ok).collect();
+
+    let entries = src_conn.query_row::<i64, _, _>(
+        "SELECT COUNT(*) FROM collection_data
+            WHERE collection_name != 'default/storage-sync-crypto'",
+        NO_PARAMS,
+        |r| r.get(0),
+    );
+    let entries_is_err = entries.is_err();
+    let entries: usize = entries.unwrap_or_default().try_into().unwrap_or_default();
+
+    let extensions = src_conn.query_row::<i64, _, _>(
+        "SELECT COUNT(DISTINCT collection_name) FROM collection_data
+            WHERE collection_name != 'default/storage-sync-crypto'",
+        NO_PARAMS,
+        |r| r.get(0),
+    );
+
+    let extensions_is_err = extensions.is_err();
+    let extensions: usize = extensions
+        .unwrap_or_default()
+        .try_into()
+        .unwrap_or_default();
+
+    // If rows.len() > entries it means we saw an error in the COUNT call (or
+    // something is very weird), but this would be caught by `entries_is_err`.
+    let entries_failed = entries.saturating_sub(rows.len());
 
-    rows.filter_map(Result::ok).collect()
+    let non_fatal_errors_seen =
+        entries_failed + (extensions_is_err as usize) + (entries_is_err as usize);
+
+    let mi = MigrationInfo {
+        entries,
+        extensions,
+        non_fatal_errors_seen,
+        // Populated later.
+        extensions_successful: 0,
+        entries_successful: 0,
+        overflowed_quotas: false,
+        open_failure: false,
+    };
+
+    (rows, mi)
 }
 
-fn do_insert(tx: &Transaction<'_>, ext_id: &str, vals: Vec<(String, Value)>) -> Result<()> {
+/// Insert the extension and values. If there are multiple values with the same
+/// key (which shouldn't be possible but who knows, database corruption causes
+/// strange things), chooses an arbitrary one. Returns the number of entries
+/// inserted, which could be different from `vals.len()` if multiple entries in
+/// `vals` have the same key.
```

Comment:
I'm a little suspect that we will want to know precisely how many corrupt databases in the wold causes sqlite to make a mistake about unique values, or what we would do even if we did. Are you sure this distinction is worthwhile?